pybio/spec/fields.py

```diff
@@ -0,0 +1,186 @@
+from importlib import import_module
+from urllib.parse import urlparse, ParseResult
+import json
+import pathlib
+import requests
+import subprocess
+import typing
+import yaml
+
+from marshmallow.fields import Str, Nested, List, Dict, Integer, Float, Tuple
+
+from pybio.exceptions import InvalidDoiException, PyBioValidationException
+from pybio.spec.pybio_types import MagicTensorsValue, MagicShapeValue
+
+
+def resolve_local_path(path_str: str, context: dict) -> pathlib.Path:
+    local_path = pathlib.Path(path_str)
+    if local_path.is_absolute():
+        return local_path.resolve()
+    elif "spec_path" in context:
+        return (context["spec_path"].parent / local_path).resolve()
+    else:
+        return local_path.absolute().resolve()
+
+
+def resolve_doi(uri: ParseResult) -> ParseResult:
+    if uri.scheme.lower() != "doi" and not uri.netloc.endswith("doi.org"):
+        return uri
+
+    doi = uri.path.strip("/")
+
+    url = "https://doi.org/api/handles/" + doi
+    r = requests.get(url).json()
+    response_code = r["responseCode"]
+    if response_code != 1:
+        raise InvalidDoiException(f"Could not resolve doi {doi} (responseCode={response_code})")
+
+    val = min(r["values"], key=lambda v: v["index"])
+
+    assert val["type"] == "URL"  # todo: handle other types
+    assert val["data"]["format"] == "string"
+    return urlparse(val["data"]["value"])
+
+
+class SpecURI(Nested):
+    # todo: improve cache location
+    cache_path = pathlib.Path(__file__).parent.parent.parent / "cache"
+
+    def _deserialize(self, value, attr, data, **kwargs):
+
+        uri = urlparse(value)
+
+        if uri.fragment:
+            raise PyBioValidationException(f"Invalid URI: {uri}. Got URI fragment: {uri.fragment}")
+        if uri.params:
+            raise PyBioValidationException(f"Invalid URI: {uri}. Got URI params: {uri.params}")
+        if uri.query:
+            raise PyBioValidationException(f"Invalid URI: {uri}. Got URI query: {uri.query}")
+
+        if uri.scheme == "file" or uri.scheme == "":
+            if uri.netloc:
+                raise PyBioValidationException(f"Invalid URI: {uri}")
+            spec_path = resolve_local_path(uri.path, self.context)
+        elif uri.netloc == "github.com":
+            orga, repo_name, blob, commit_id, *in_repo_path = uri.path.strip("/").split("/")
+            in_repo_path = "/".join(in_repo_path)
+            cached_repo_path = self.cache_path / orga / repo_name / commit_id
+            spec_path = cached_repo_path / in_repo_path
+            if not spec_path.exists():
+                cached_repo_path = cached_repo_path.resolve().as_posix()
```

Comment:
yes
```python
from pathlib import Path
a = Path("/lala/here and there/lülö-da")
a
Out[4]: WindowsPath('/lala/here and there/lülö-da')
a.as_posix()
Out[5]: '/lala/here and there/lülö-da'
Path(a.as_posix())
Out[6]: WindowsPath('/lala/here and there/lülö-da')
```

Reply:
Sorry, my initial comment was not quite clear. 
I was worrying if this works in the `subprocess.call` to `git`, because it would be something like this:
```
git clone reponame.git /path/to/repo
```
Where `/path/to/repo` is encoded as posix path, which might cause problems on Windows?